internal/memory/backend.go

```diff
@@ -0,0 +1,189 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memory
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"buf.build/gen/go/lekkodev/cli/bufbuild/connect-go/lekko/backend/v1beta1/backendv1beta1connect"
+	backendv1beta1 "buf.build/gen/go/lekkodev/cli/protocolbuffers/go/lekko/backend/v1beta1"
+	"github.com/bufbuild/connect-go"
+	"github.com/cenkalti/backoff/v4"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+const (
+	lekkoAPIKeyHeader = "apikey"
+)
+
+type Store interface {
+	Evaluate(key string, namespace string, lc map[string]interface{}, dest proto.Message) error
+	Close(ctx context.Context) error
+}
+
+// Constructs an in-memory store that fetches configs from lekko's backend.
+func NewBackendStore(ctx context.Context, apikey, url, ownerName, repoName string, updateInterval time.Duration) (Store, error) {
+	return newBackendStore(ctx, apikey, ownerName, repoName, updateInterval, backendv1beta1connect.NewDistributionServiceClient(http.DefaultClient, url))
+}
+
+func newBackendStore(ctx context.Context, apikey, ownerName, repoName string, updateInterval time.Duration, distClient backendv1beta1connect.DistributionServiceClient) (*backendStore, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	b := &backendStore{
+		distClient: distClient,
+		store:      newStore(),
+		repoKey: &backendv1beta1.RepositoryKey{
+			OwnerName: ownerName,
+			RepoName:  repoName,
+		},
+		apikey:         apikey,
+		updateInterval: updateInterval,
+		cancel:         cancel,
+	}
+	// register with lekko backend
+	if err := b.registerWithBackoff(ctx); err != nil {
+		return nil, errors.Wrap(err, "error registering client")
+	}
+	// initialize the store once with configs
+	if _, err := b.updateStoreWithBackoff(ctx); err != nil {
+		return nil, err
+	}
+	// kick off an asynchronous goroutine that updates the store periodically
+	b.loop(ctx)
+	return b, nil
+}
+
+type backendStore struct {
+	distClient         backendv1beta1connect.DistributionServiceClient
+	store              *store
+	repoKey            *backendv1beta1.RepositoryKey
+	apikey, sessionKey string
+	wg                 sync.WaitGroup
+	updateInterval     time.Duration
+	cancel             context.CancelFunc
+}
+
+// Close implements Store.
+func (b *backendStore) Close(ctx context.Context) error {
+	// cancel any ongoing background loops
+	b.cancel()
+	// wait for background work to complete
+	b.wg.Wait()
+	_, err := b.distClient.DeregisterClient(ctx, connect.NewRequest(&backendv1beta1.DeregisterClientRequest{
+		SessionKey: b.sessionKey,
+	}))
+	return err
+}
+
+// Evaluate implements Store.
+func (b *backendStore) Evaluate(key string, namespace string, lc map[string]interface{}, dest protoreflect.ProtoMessage) error {
```

Comment:
what's `lc`?

Reply:
lekko context, since context is already taken by go :) i can expand it to make it more clear